gapir/client/client.go

```diff
@@ -32,84 +37,195 @@ type tyLaunchArgsKey string
 // line arguments when launching GAPIR. The property must be of type []string.
 const LaunchArgsKey tyLaunchArgsKey = "gapir-launch-args"
 
+type clientInfo struct {
+	device               bind.Device
+	arch                 device.Architecture
+	abi                  *device.ABI
+	deviceConnectionInfo deviceConnectionInfo
+	connection           gapir.Connection
+	bgConnection         *backgroundConnection
+}
+
+type deviceArch struct {
+	device bind.Device
+	arch   device.Architecture
+}
+
+// ConnectionKey is used by manager to obtain a connection
+type ConnectionKey deviceArch
+
 // Client is interface used to connect to GAPIR instances on devices.
 type Client struct {
-	mutex    sync.Mutex
-	sessions map[deviceArch]*session
+	mutex       sync.Mutex
+	clientInfos map[ConnectionKey]clientInfo
 }
 
+const (
+	connectTimeout    = time.Second * 10
+	heartbeatInterval = time.Millisecond * 500
+)
+
 // New returns a newly construct Client.
 func New(ctx context.Context) *Client {
-	c := &Client{sessions: map[deviceArch]*session{}}
+	client := &Client{clientInfos: map[ConnectionKey]clientInfo{}}
 	app.AddCleanup(ctx, func() {
-		c.shutdown(ctx)
+		client.shutdown(ctx)
 	})
-	return c
-}
-
-type deviceArch struct {
-	d bind.Device
-	a device.Architecture
+	return client
 }
 
 // Connect opens a connection to the replay device.
-func (c *Client) Connect(ctx context.Context, d bind.Device, abi *device.ABI) (gapir.Connection, error) {
+func (client *Client) Connect(ctx context.Context, device bind.Device, abi *device.ABI) (*ConnectionKey, error) {
+	client.mutex.Lock()
+	defer client.mutex.Unlock()
+
 	ctx = status.Start(ctx, "Connect")
 	defer status.Finish(ctx)
 
-	s, isNew, err := c.getOrCreateSession(ctx, d, abi)
+	if client.clientInfos == nil {
+		return nil, log.Err(ctx, nil, "Client has been shutdown")
+	}
+
+	deviceArch := deviceArch{device: device, arch: abi.GetArchitecture()}
+	key := ConnectionKey(deviceArch)
+
+	if clientInfo, ok := client.clientInfos[key]; ok {
+		if clientInfo.abi.SameAs(abi) {
+			return &key, nil
+		}
+
+		// TODO: Should we reconnect instead of disconnect?
+		client.closeConnection(ctx, key)
```

Comment:
The `key` depends on the ABI, so the same device with two different ABI should lead to two different keys.
If this is correct, then your test here is probably not useful, as the same key cannot lead to different ABIs. In other words, `clientInfo.abi.SameAs(abi)` is always true. If so, maybe we should check that the connection already registered in `client.clientInfos[key].connection` is still valid or not. If it is, just return `&key`, if not, then re-create it.

Reply:
You are right it's a leftover from the old code. For the second part, if the connection is invalid, it's being recreated. So, if a connection exists then it's valid.